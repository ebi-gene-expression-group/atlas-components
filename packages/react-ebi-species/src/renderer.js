"use strict";

//*------------------------------------------------------------------*

var React = require('react');
var ReactDOM = require('react-dom');

//*------------------------------------------------------------------*

var Icon = require('./SpeciesIcon.jsx');

//*------------------------------------------------------------------*

module.exports = function(mountNode,species,colourOverride, colourPerKingdom) {
    ReactDOM.render(
        React.createElement(Icon,{
          species: species,
          colourOverride: colourOverride,
          colourPerKingdom: colourPerKingdom
        }), mountNode
    );
};
