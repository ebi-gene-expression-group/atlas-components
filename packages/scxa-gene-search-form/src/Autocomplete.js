import React from 'react'
import PropTypes from 'prop-types'
import URI from 'urijs'

import AsyncCreatableSelect from 'react-select/lib/AsyncCreatable'
import ebiVfReactSelectReplacements from './ebiVfReactSelectReplacements'

const _asyncFetchOptions = (atlasUrl, suggesterEndpoint, selectedSpecies, allSpecies) =>
  async (inputValue) => {
    const suggesterUrl = URI(suggesterEndpoint, atlasUrl).search({
      query: inputValue,
      species: selectedSpecies || allSpecies.join(`,`)
    }).toString()

    const response = await fetch(suggesterUrl)
    if (response.ok) {
      return await response.json()
    }
    throw new Error(`${suggesterUrl} => ${response.status}`)
  }

const Autocomplete = ({atlasUrl, suggesterEndpoint, selectedSpecies, allSpecies, onChange, defaultValue}) => {
  const _defaultValue = defaultValue.term && defaultValue.term.trim() ?
    {
      label: defaultValue.term.trim(),
      value: defaultValue.category && defaultValue.category.trim() ?
        JSON.stringify(defaultValue) :
        JSON.stringify({term: defaultValue.term.trim(), category: `q`})
    } :
    null

  return (
    <div>
      <label key={`label`} htmlFor={`geneQuery`}>Gene ID or gene symbol</label>
      <AsyncCreatableSelect
        key={`autocomplete`}
        name={`geneQuery`}
        components={{ DropdownIndicator: null, IndicatorSeparator: null }}
        styles={ebiVfReactSelectReplacements.styles}
        onChange={onChange}
        loadOptions={_asyncFetchOptions(atlasUrl, suggesterEndpoint, selectedSpecies, allSpecies)}
        noOptionsMessage={() => null}
        allowCreateWhileLoading={true}
        isClearable={true}
        createOptionPosition={`first`}
        formatCreateLabel={(inputValue) => inputValue}
        isValidNewOption={(inputValue) => inputValue.trim() !== ``}
        getNewOptionData={(inputValue) => ({
          label: inputValue,
          value: JSON.stringify({term: inputValue, category: `q`})
        })}
        placeholder={``}
        defaultValue={_defaultValue}/>
    </div>
  )
}

Autocomplete.propTypes = {
  atlasUrl: PropTypes.string.isRequired,
  suggesterEndpoint: PropTypes.string.isRequired,
  selectedSpecies: PropTypes.string,
  allSpecies: PropTypes.arrayOf(PropTypes.string),
  onChange: PropTypes.func.isRequired,
  defaultValue: PropTypes.shape({
    term: PropTypes.string,
    category: PropTypes.string
  })
}

Autocomplete.defaultProps = {
  allSpecies: [],
  defaultValue: {}
}

export default Autocomplete
