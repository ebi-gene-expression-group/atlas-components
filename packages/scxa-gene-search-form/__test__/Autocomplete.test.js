import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import fetchMock from 'fetch-mock'
import URI from 'urijs'

import '@babel/polyfill'
import Autocomplete from '../src/Autocomplete'
import AsyncCreatableSelect from 'react-select/async-creatable'

const getRandomInt = (max) => Math.floor(Math.random() * Math.floor(max))

const props = {
  atlasUrl: `foo/`,
  suggesterEndpoint: `suggest`,
  onChange: () => {}
}

const defaultValue = {
  term: `foo`,
  category: `bar`
}

const species = [
  `Meeseek`,
  `Gromflomite`,
  `Cromulon`,
  `Zigerion`,
  `Moopian`,
  `Bliznarvian`,
  `Greebybobe`
]

describe(`Autocomplete`, () => {
  test(`displays the default value`, () => {
    const wrapper = shallow(<Autocomplete {...props} defaultValue={defaultValue} />)
    expect(wrapper.find(AsyncCreatableSelect).props().defaultValue).toEqual({
      label: defaultValue.term,
      value: JSON.stringify(defaultValue)
    })
  })

  test(`matches snapshot`, () => {
    const tree =
      renderer
        .create(<Autocomplete Autocomplete {...props} defaultValue={defaultValue} allSpecies={species} />)
        .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe(`Autocomplete suggestions fetch`, () => {
  beforeEach(() => {
    fetchMock.restore()
  })

  test(`calls suggester with selected species as request parameter`, () => {
    fetchMock.get(`*`, [])

    const randomSpecies = species[getRandomInt(species.length)]
    const wrapper =
      shallow(<Autocomplete {...props} selectedSpecies={randomSpecies} />)
        .find(AsyncCreatableSelect).dive()

    wrapper.simulate(`inputChange`, `yeah`)

    expect(fetchMock.calls()).toHaveLength(1)
    expect(URI(fetchMock.lastUrl()).search(true)).toHaveProperty(`species`, randomSpecies)
  })

  test(`calls suggester with all species as request parameter if no selected species is provided`, () => {
    fetchMock.get(`*`, [])

    const wrapper =
      shallow(<Autocomplete {...props} allSpecies={species} />)
        .find(AsyncCreatableSelect).dive()

    wrapper.simulate(`inputChange`, `yeah`)

    expect(fetchMock.calls()).toHaveLength(1)
    expect(URI(fetchMock.lastUrl()).search(true)).toHaveProperty(`species`, species.join(`,`))

  })
})
