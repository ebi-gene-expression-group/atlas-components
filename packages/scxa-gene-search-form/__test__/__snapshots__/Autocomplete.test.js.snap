// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete matches snapshot 1`] = `
<div>
  <label
    htmlFor="geneQuery"
  >
    Gene ID or gene symbol
  </label>
  <div
    className=" css-2b097c-container"
    onKeyDown={[Function]}
  >
    <div
      className=" css-tinoic"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className=" css-1hwfws3"
      >
        <div
          className=" css-1uccc91-singleValue"
        >
          foo
        </div>
        <div
          className="css-1g6gooi"
        >
          <div
            className=""
            style={
              Object {
                "display": "inline-block",
              }
            }
          >
            <input
              aria-autocomplete="list"
              autoCapitalize="none"
              autoComplete="off"
              autoCorrect="off"
              disabled={false}
              id="gene-input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              spellCheck="false"
              style={
                Object {
                  "background": 0,
                  "border": 0,
                  "boxSizing": "content-box",
                  "color": "inherit",
                  "fontSize": "inherit",
                  "label": "input",
                  "opacity": 1,
                  "outline": 0,
                  "padding": 0,
                  "width": "1px",
                }
              }
              tabIndex="0"
              type="text"
              value=""
            />
            <div
              style={
                Object {
                  "height": 0,
                  "left": 0,
                  "overflow": "scroll",
                  "position": "absolute",
                  "top": 0,
                  "visibility": "hidden",
                  "whiteSpace": "pre",
                }
              }
            >
              
            </div>
          </div>
        </div>
      </div>
      <div
        className=" css-1wy0on6"
      >
        <div
          aria-hidden="true"
          className=" css-tlfecz-indicatorContainer"
          onMouseDown={[Function]}
          onTouchEnd={[Function]}
        >
          <svg
            aria-hidden="true"
            className="css-19bqh2r"
            focusable="false"
            height={20}
            viewBox="0 0 20 20"
            width={20}
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
      </div>
    </div>
    <input
      name="geneQuery"
      type="hidden"
      value="{\\"term\\":\\"foo\\",\\"category\\":\\"bar\\"}"
    />
  </div>
</div>
`;
